cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/utilities/assert.cuh>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/equal.h>
+#include <thrust/swap.h>
+#include <thrust/transform_reduce.h>
+
+#include <cuda/std/tuple>
+#include <cuda/std/utility>
+
+#include <limits>
+#include <memory>
+#include <utility>
+
+namespace cudf {
+namespace experimental {
+
+template <cudf::type_id t>
+struct non_nested_id_to_type {
+  using type = std::conditional_t<cudf::is_nested(data_type(t)), void, id_to_type<t>>;
+};
+
+namespace lexicographic_comparison {
+
+/**
+ * @brief Performs a relational comparison between two elements in two columns.
+ *
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <typename Nullate>
+class element_relational_comparator {
```

Comment:
`cudf::experimental::lexicographical_comparison::element_relational_comparator` is a bit of a mouthful :) 

We don't have to resolve this in this particular PR, but I feel like there's a more concise namespace/type name scheme here.

Maybe something like:
```
namespace cudf::detail::row::lex{
   class element_comparator;
   class row_comparator;
   class self_row_comparator;
   class preprocessed_table;
}
```

And then we'd have `cudf::detail::row::eq` and `cudf::detail::row::hash` namespaces? Maybe `eq` and `lex` are too abbreviated? ¯\\\_(ツ)_/¯

Reply:
I love `eq` and `lex` but so far. I've seen abbreviations are discouraged in libcudf.

Reply:
I make it `cudf::experimental::lex::self_comparator`. Do you think the `row` namespace it essential to convey the `lex` meaning? Because `row` namespace will also contain `element_comparator`.

Reply:
I feel like `row` should be in the name somewhere if not the namespace. Though I do still like the `row` namespace as a way to group the row comparators and the row hasher. 

> Because row namespace will also contain element_comparator

This doesn't bother me much as the `element_comparator` is really just an implementation detail of the `row_comparator` and not a utility usable on it's own. In fact, we should consider moving the `element_comparator` to be a private nested type of the `row_comparator` to make it clear it's not externally usable. And with C++20, it doesn't even need to be a type, but can be a template lambda...

Reply:
new namespace:
```
namespace experimental::row::lexicographic{
  self_comparator()
  comparator() // future
}
```
And when this leaves experimental in the future, it will read `row::lexicographic::comparator` which is nice because it looks so much like the old one.